File: src/main/java/cn/edu/jxnu/base/controller/admin/system/ResourceController.java
/* 梦境迷离 (C)2020 */
package cn.edu.jxnu.base.controller.admin.system;

import cn.edu.jxnu.base.controller.BaseController;
import cn.edu.jxnu.base.entity.Resource;
import cn.edu.jxnu.base.entity.ZtreeView;
import cn.edu.jxnu.base.service.IResourceService;
import cn.edu.jxnu.base.service.specification.SimpleSpecificationBuilder;
import cn.edu.jxnu.base.service.specification.SpecificationOperator.Operator;
import cn.edu.jxnu.base.utils.JsonResult;
import javax.servlet.http.HttpServletRequest;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.ResponseBody;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * 系统资源控制类
 *
 * @author 梦境迷离
 * @version V2.0 2020年11月20日
 */
@Controller
@RequestMapping("/admin/resource")
public class ResourceController extends BaseController {

  @Autowired private IResourceService resourceService;

  /**
   * 资源树
   *
   * @param resourceId 资源ID
   * @return Flux ZtreeView
   */
  @RequestMapping("/tree/{resourceId}")
  @ResponseBody
  public Flux<ZtreeView> tree(@PathVariable Integer resourceId) {
    return resourceService.tree(resourceId);
  }

  /**
   * 打开资源管理首页
   *
   * @return String
   */
  @RequestMapping("/index")
  public String index() {
    return "admin/resource/index";
  }

  /**
   * 资源管理分页
   *
   * @param request request
   * @return Mono Page
   */
  @RequestMapping(
      value = "/list",
      method = {RequestMethod.POST})
  @ResponseBody
  public Mono<Page<Resource>> list(HttpServletRequest request) {
    SimpleSpecificationBuilder<Resource> builder = new SimpleSpecificationBuilder<>();
    String searchText = request.getParameter("searchText");
    if (StringUtils.isNotBlank(searchText)) {
      builder.add("name", Operator.likeAll.name(), searchText);
    }
    return resourceService.findAll(builder.generateSpecification(), getPageRequest(request)).log();
  }

  /**
   * 打开资源添加页面
   *
   * @param map map
   * @return String
   */
  @RequestMapping(value = "/add")
  public String add(ModelMap map) {
    Flux<Resource> list = resourceService.findAll();
    list.log().subscribe(l -> map.put("list", l));
    return "admin/resource/form";
  }

  /**
   * 打开资源修改页面
   *
   * @param id 资源ID
   * @param map map
   * @return String
   */
  @RequestMapping(value = "/edit/{id}")
  public String edit(@PathVariable Integer id, ModelMap map) {
    resourceService
        .find(id)
        .map(
            r -> {
              map.put("resource", r);
              return resourceService.findAll().collectList().map(l -> map.put("list", l));
            })
        .log()
        .subscribe();
    return "admin/resource/form";
  }

  /**
   * 资源添加或修改
   *
   * @param resource 资源
   * @return Mono JsonResult
   */
  @RequestMapping(
      value = {"/edit"},
      method = {RequestMethod.POST})
  @ResponseBody
  public Mono<JsonResult> edit(Resource resource) {
    // 可能存在需要在增加资源的时候默认给管理员增加权限
    return resourceService
        .saveOrUpdate(resource)
        .map(r -> JsonResult.success())
        .log()
        .onErrorResume(error -> Mono.just(JsonResult.failure(error.getLocalizedMessage())));
  }

  /**
   * 资源删除
   *
   * @param id 资源ID
   * @return Mono JsonResult
   */
  @RequestMapping(
      value = "/delete/{id}",
      method = {RequestMethod.POST})
  @ResponseBody
  public Mono<JsonResult> delete(@PathVariable Integer id) {
    return resourceService
        .delete(id)
        .map(r -> r ? JsonResult.success() : JsonResult.failure("删除失败"))
        .log()
        .onErrorResume(error -> Mono.just(JsonResult.failure(error.getLocalizedMessage())));
  }
}

File: src/main/java/cn/edu/jxnu/base/controller/admin/system/ResourceController.scala
package cn.edu.jxnu.base.controller.admin.system

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.beans.factory.annotation.Autowired
import cn.edu.jxnu.base.service.IResourceService
import cn.edu.jxnu.base.controller.BaseController
import org.springframework.web.bind.annotation.ResponseBody
import org.springframework.web.bind.annotation.PathVariable
import cn.edu.jxnu.base.entity.ZtreeView
import cn.edu.jxnu.base.service.specification.SimpleSpecificationBuilder
import org.apache.commons.lang3.StringUtils
import cn.edu.jxnu.base.entity.Resource
import org.springframework.data.domain.Page
import cn.edu.jxnu.base.service.specification.SpecificationOperator.Operator
import org.springframework.ui.ModelMap
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import cn.edu.jxnu.base.common.JsonResult
import scala.collection.immutable.{ List => _, _ }
import java.util.{ List => JavaList }

/**
 *
 * 系统资源控制类
 *
 * @author 梦境迷离.
 * @time 2018年5月17日
 * @version v2.0
 */
@Controller
@RequestMapping(value = { Array("/admin/resource") })
class ResourceController @Autowired() (resourceService: IResourceService) extends BaseController {

    @RequestMapping(value = { Array("/tree/{resourceId}") })
    @ResponseBody
    def tree(@PathVariable resourceId: Int): JavaList[ZtreeView] = {
        resourceService.tree(resourceId)

    }

    @RequestMapping(value = { Array("/index") })
    def index() = {
        "admin/resource/index"
    }

    @RequestMapping(value = { Array("/list") })
    @ResponseBody
    def list(): Page[Resource] = {
        val builder = new SimpleSpecificationBuilder[Resource]()
        val searchText = request.getParameter("searchText")
        if (StringUtils.isNotBlank(searchText)) {
            builder.add("name", Operator.likeAll.name(), searchText)
        }
        resourceService.findAll(builder.generateSpecification(), getPageRequest())
    }

    @GetMapping(value = { Array("/add") })
    def add(map: ModelMap) = {
        val list = resourceService.findAll()
        map.put("list", list)
        "admin/resource/form"
    }

    @GetMapping(value = { Array("/edit/{id}") })
    def edit(@PathVariable id: Int, map: ModelMap) = {
        val resource = resourceService.find(id)
        map.put("resource", resource)
        val list = resourceService.findAll()
        map.put("list", list)
        "admin/resource/form"
    }

    @ResponseBody
    @PostMapping(value = { Array("/edit") })
    def edit(resource: Resource, map: ModelMap): JsonResult = {
        try {
            resourceService.saveOrUpdate(resource)
            // 可能存在需要在增加资源的时候默认给管理员增加权限
        } catch {
            case e: Exception =>
                return JsonResult.failure(e.getMessage())
        }
        JsonResult.success()
    }

    @PostMapping(value = { Array("/delete/{id}") })
    @ResponseBody
    def delete(@PathVariable id: Int, map: ModelMap): JsonResult = {
        try {
            resourceService.delete(id)
        } catch {
            case e: Exception =>
                e.printStackTrace()
                return JsonResult.failure(e.getMessage())
        }
        JsonResult.success()
    }
}